src/decode/zero.gleam

```diff
@@ -0,0 +1,846 @@
+//// The `Dynamic` type is used to represent dynamically typed data. That is, data
+//// that we don't know the precise type of yet, so we need to introspect the data to
+//// see if it is of the desired type before we can use it. Typically data like this
+//// would come from user input or from untyped languages such as Erlang or JavaScript.
+////
+//// This module provides the `Decoder` type and associated functions, which provides
+//// a type-safe and composable way to convert dynamic data into some desired type,
+//// or into errors if the data doesn't have the desired structure.
+////
+//// The `Decoder` type is generic and has 1 type parameter, which is the type that
+//// it attempts to zero. A `Decoder(String)` can be used to decode strings, and a
+//// `Decoder(Option(Int))` can be used to decode `Option(Int)`s
+////
+//// Decoders work using _runtime reflection_ and the data structures of the target
+//// platform. Differences between Erlang and JavaScript data structures may impact
+//// your decoders, so it is important to test your decoders on all supported
+//// platforms.
+////
+//// The decoding technique used by this module was inspired by Juraj Petráš'
+//// [Toy](https://github.com/Hackder/toy), Go's `encoding/json`, and Elm's
+//// `Json.Decode`. Thank you to them!
+////
+//// # Examples
+////
+//// Dynamic data may come from various sources and so many different syntaxes could
+//// be used to describe or construct them. In these examples a pseudocode
+//// syntax is used to describe the data.
+////
+//// ## Simple types
+////
+//// This module defines decoders for simple data types such as [`string`](#string),
+//// [`int`](#int), [`float`](#float), [`bit_array`](#bit_array), and [`bool`](#bool).
+////
+//// ```gleam
+//// // Data:
+//// // "Hello, Joe!"
+////
+//// let result = zero.run(data, zero.string)
+//// assert result == Ok("Hello, Joe!")
+//// ```
+////
+//// ## Lists
+////
+//// The [`list`](#list) decoder decodes `List`s. To use it you must construct it by
+//// passing in another decoder into the `list` function, which is the decoder that
+//// is to be used for the elements of the list, type checking both the list and its
+//// elements.
+////
+//// ```gleam
+//// // Data:
+//// // [1, 2, 3, 4]
+////
+//// let result = zero.run(data, zero.list(zero.int))
+//// assert result == Ok([1, 2, 3, 4])
+//// ```
+////
+//// On Erlang this decoder can decode from lists, and on JavaScript it can
+//// decode from lists as well as JavaScript arrays.
```

Comment:
is this accurate for JS?  might be missing something, but it looks like the FFI for this checks if `instanceof List` and fails otherwise

Reply:
Good catch, thank you.